### PR TITLE
Fix min_by and max_by aggregate functions for DATE inputs

### PR DESCRIPTION
**Pull request by pramodsatya**
The min and max bounds used in min_by, max_by aggregate functions for DATE
datatype are both initialized to zero. This incorrect initialization is fixed in this PR
and the min, max bounds are initialized to minimum and maximum values 
possible for DATE respectively.

### CODE REVIEW

**Comment by pedroerp** [comment]
@xiaoxmeng could you help review this PR since you recently changed this function?

**Review by aditi-pandit — COMMENTED** [review]
Lets try a different approach as discussed where isNumeric and isDate are separated for making the code more clean.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 1]
```diff
@@ -998,5 +998,17 @@ VELOX_INSTANTIATE_TEST_SUITE_P(
     MinMaxByGroupByAggregationTest,
     testing::ValuesIn(getTestParams()));
 
+TEST_F(MinMaxByGroupByAggregationTest, largeValuesDate) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:1001]: Like we discussed we should change this to use a similar framework for the other type testing in this file.

Changing valueAt for Date to return the corresponding string for date should suffice.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 2]
```diff
@@ -23,6 +23,21 @@
 
 using namespace facebook::velox::exec::test;
 using facebook::velox::VectorFuzzer;
+
+template <>
+struct fmt::formatter<facebook::velox::Date> {
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext& ctx) {
+    return ctx.begin();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:31]: What does this do ? Is no special logic needed here ?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 3]
```diff
@@ -955,5 +984,17 @@ VELOX_INSTANTIATE_TEST_SUITE_P(
     MinMaxByGroupByAggregationTest,
     testing::ValuesIn(getTestParams()));
 
+TEST_F(MinMaxByGroupByAggregationTest, largeValuesDate) {
+  auto vectors = {makeRowVector(
+      {makeConstant(Date(60577), 100), makeConstant(Date(-57604), 100)})};
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:989]: Please write something like 

auto parseDate = [](const std::string& dateStr) {
    Date returnDate;
    parseTo(dateStr, returnDate);
    return returnDate;
  };

and use this function to give the string representations of the dates you are using in the tests.

Comment (pramodsatya) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp]: Thanks for the suggestion, this test case is now removed and Date() tests are integrated with the rest of the aggregation test suite.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 2 above)
Comment (pramodsatya) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:31]: No special logic is needed since just the string formatter is used here. I followed the approach from: https://wgml.pl/blog/formatting-user-defined-types-fmt.html.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 4]
```diff
@@ -97,15 +102,38 @@ template <typename T, typename = void>
...
 
 template <typename T>
-struct AccumulatorTypeTraits<T, std::enable_if_t<!isNumeric<T>(), void>> {
+struct AccumulatorTypeTraits<
+    T,
+    std::enable_if_t<!isNumeric<T>() and !isDate<T>(), void>> {
   using AccumulatorType = SingleValueAccumulator;
 };
 
+template <typename T>
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:118]: Move this code above, just immediately below the case for isNumeric at line 110 for better readability.

**Review by aditi-pandit — COMMENTED** [review]
Looks good. We can open draft for external review.

**Comment by mbasmanova** [comment]
@pramodsatya Thank you for the fix. CI is failing. Would you take a look?

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 5]
```diff
@@ -97,15 +102,38 @@ template <typename T, typename = void>
...
 template <typename T>
-struct AccumulatorTypeTraits<T, std::enable_if_t<isNumeric<T>(), void>> {
+struct AccumulatorTypeTraits<
+    T,
+    std::enable_if_t<isNumeric<T>() and !isDate<T>(), void>> {
+  using AccumulatorType = T;
+};
+
+template <typename T>
+struct AccumulatorTypeTraits<T, std::enable_if_t<isDate<T>(), void>> {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:112]: This struct looks the same as the one for numeric types. Can these be combined?

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 6]
```diff
@@ -32,13 +32,18 @@ constexpr bool isNumeric() {
...
+}
+
 template <typename T, typename TAccumulator>
 void extract(
     TAccumulator* accumulator,
     const VectorPtr& vector,
     vector_size_t index,
     T* rawValues) {
-  if constexpr (isNumeric<T>()) {
+  if constexpr (isNumeric<T>() or isDate<T>()) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:46]: Looks like isNumeric and isDate are always used together. Consider replacing these with a single function isNumericOrDate.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 7]
```diff
@@ -23,6 +23,21 @@
 
 using namespace facebook::velox::exec::test;
 using facebook::velox::VectorFuzzer;
+
+template <>
+struct fmt::formatter<facebook::velox::Date> {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:28]: If we want to add support for Date type to fmt::format, this code should go somewhere else where it will be available to anyone using the type, e.g. Date.h. That change should be in its own PR with a test.

**Review by mbasmanova — COMMENTED** [review]
@pramodsatya Looks good overall % some comments below. The PR description says "Fix initialization", but doesn't explain what was the problem with the initialization and what is the solution. Please, update.

**Comment by mbasmanova** [comment]
@pramodsatya Pramod, thank you for updating the PR description. It is much easier to understand the changes now. Would you also format the PR description to wrap at 80 characters? BTW, do we now have tests that fail without the fix?

**Comment by pramodsatya** [comment]
Sure, will update the description, thanks! Currently, we do not have tests that fail without the fix, the tests for DATE have also been added in this PR and these tests do fail without the fix.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp (hunk 6 above)
Comment (pramodsatya) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp]: Thanks for the suggestion, reverted back to single function isNumericOrDate.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 7 above)
Comment (pramodsatya) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp]: Thanks for the suggestion, added formatter changes in PR #2457, will revert the Date.h changes containing the custom formatter after #2457 is checked in.

**Review by mbasmanova — APPROVED** [review]
@pramodsatya Looks good. Ping me once the fmt::format PR has landed and this one is rebased. I'll merge this one then.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 8]
```diff
@@ -115,7 +122,7 @@ std::vector<TestParam> getTestParams() {
 
 class MinMaxByAggregationTestBase : public AggregationTestBase {
  protected:
-  MinMaxByAggregationTestBase() : numValues_(6) {}
+  MinMaxByAggregationTestBase() : numValues_(7) {}
```
Comment (bikramSingh91) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:125]: just curious about the change of numValues_ as to why it was moved to 7.  I wonder if the tests for date only failed after the move to 7 values.

Comment (pramodsatya) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:125]: Thanks for pointing it out, upon review I see this change wasn't needed, but the tests for date failed without this change. Would it be fine if numValues_ remains 7 or would you suggest reverting it back to 6? It should be fine either way.

Comment (bikramSingh91) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:125]: Thanks for checking, should be fine if we leave it as 7.
Only reason I asked was that I was just curious if there was more to it, and if so would have suggested to document it.

Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:125]: > upon review I see this change wasn't needed, but the tests for date failed without this change

What is the failure?

Comment (pramodsatya) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:125]: Sorry for the confusion, I meant the min_by, max_by tests for date are not failing because of this change. The failure is because of the incorrect initialization as updated in the PR description.

Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:125]: @pramodsatya I'm trying to understand why we need numValues_ to be 7 and not 6? Can it be 6? If so, let's revert the change to 7 to avoid confusion.

Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp]: Thank you, @pramodsatya .
